crmsh/corosync.py

```diff
@@ -512,3 +514,309 @@ def remove_key(cls, path, index=0):
+        Args:
+            config: [in/out] the configuration dom
+            links: [in] parsed link data
+            linknumber: [in] the linknunmber to add or update
+            node_addresses: [in] a mapping from nodeid to node address.
+
+        Returns:
+            a reference to in/out arg `config`
+        """
+        existing_addr_node_map = {
+            utils.IP(node.addr).ip_address: node.nodeid
+            for link in links if link is not None
+                for node in link.nodes
+            if node.addr != ''
+        }
+        for nodeid, addr in node_addresses.items():
+            found = next((node for node in links[linknumber].nodes if node.nodeid == nodeid), None)
+            if found is None:
+                raise ValueError(f'Unknown nodeid {nodeid}.')
+            canonical_addr = utils.IP(addr).ip_address
+            if (
+                    found.addr == ''    # adding a new addr
+                    or utils.IP(found.addr).ip_address != canonical_addr    # updating a addr and the new value is not the same as the old value
+            ):
+                # need to change uniqueness
+                existing = existing_addr_node_map.get(canonical_addr, None)
+                if existing is not None:
+                    raise LinkManager.DuplicatedNodeAddressException(addr, nodeid, existing)
+            found.addr = addr
+            existing_addr_node_map[canonical_addr] = found.nodeid
+        nodes = config['nodelist']['node']
+        assert isinstance(nodes, list)
+        for node in nodes:
+            updated_addr = node_addresses.get(int(node['nodeid']), None)
+            if updated_addr is not None:
+                node[f'ring{linknumber}_addr'] = updated_addr
+        return config
+
+    def add_link(self, node_addresses: typing.Mapping[int, str], options: dict[str, str|None]) -> dict:
+        links = self.links()
+        next_linknumber = next((i for i, link in enumerate(links) if link is None), -1)
+        if next_linknumber == -1:
+            raise ValueError(f'Cannot add a new link. The maximum number of links supported is {KNET_LINK_NUM_LIMIT}.')
+        nodes = next(x for x in links if x is not None).nodes
+        unspecified_nodes = [node.nodeid for node in nodes if node.nodeid not in node_addresses]
+        if unspecified_nodes:
+            raise self.MissingNodesException(unspecified_nodes)
+        links[next_linknumber] = Link(next_linknumber, [dataclasses.replace(node, addr='') for node in nodes])
+        self.__upsert_node_addr_impl(self._config, links, next_linknumber, node_addresses)
+        return self.update_link(next_linknumber, options)
+
+    def remove_link(self, linknumber: int) -> dict:
+        """Remove the specified link.
+
+        Parameters:
+            * linknumber: the link to update
+        Returns: updated configuration dom. The internal state of LinkManager is also updated.
+        """
+        links = self.links()
```

Comment:
To support `crm corosync link remove -1`

```suggestion
        links = self.links()
        if linknumber == -1:
            for i in reversed(range(8)):
                logger.debug(f"ROGER: links[{i}] = {links[i]}")
                if links[i]:
                    linknumber = i
                    break

```